### ISSUE TITLE
The toolbar icon becomes too large under HiDPI from 0.6.0

### ISSUE THREAD

**Issue by Hexcles** [akhodakivskiy/VimFx#591]
I believe the fix for Issue #383 (probably the commit 3552282) introduced some compatibility issues with [AutoHiDPI](https://addons.mozilla.org/firefox/addon/autohidpi/), which is a simple plugin to tweak `layout.css.devPixelsPerPx` according to the screen resolution.

It seems the high-resolution version of the icon is now included, but under HiDPI (i.e. when `layout.css.devPixelsPerPx` is larger than 1), the high-resolution icon is scaled again, leading to a too large and still blurry icon. See the screenshot:
![screenshot](https://i.imgur.com/uXgzk9l.png)

Furthermore, it seems to me that AutoHiDPI (and the `layout.css.devPixelsPerPx` setting) is not at fault. As you can see from the screenshot, [uBlock Origin](https://github.com/gorhill/uBlock) is using a high-resolution icon and showing it at the correct size. I believe you would find some inspiration from its source code.

**Comment by lydell** [comment]
Thanks for reporting! I don’t have a HiDPI screen to test on, so I simply copy-pasted the [suggested code on MDN](https://developer.mozilla.org/en-US/Firefox/Australis_add-on_compat).

Help on this would be more than welcome. This should be really easy to contribute even for a non-programmer, since there are only CSS changes needed. [Here’s the code currently used.](https://developer.mozilla.org/en-US/Firefox/Australis_add-on_compat)

Alternatively, you could try to fix it with custom [styling](https://github.com/akhodakivskiy/VimFx/blob/master/documentation/styling.md) and report back what code works.

**Comment by lydell** [comment]
I found uBlock Origin's source code difficult to understand. My go-to choice of source code inspiration is Adblock Plus; and its source code was easier to understand in this case. Would you mind checking if Adblock Plus' button looks correct on your system? If so I could copy their approach.

**Comment by lydell** [comment]
Just installed uBlock Origin and inspected its toolbar button. They use an SVG icon. SVG:s look perfect in any size. Unfortunately, VimFx’s icon is a PNG (in different sizes). So looking at uBlock Origin is of no help.

**Comment by lydell** [comment]
I think I figured out what was wrong. Never mind the other comments. Thanks for such a well-written bug report!

**Comment by Hexcles** [comment]
Thanks for your quick response! And I confirm this issue is fixed in 0.6.2.

**Comment by dwhinham** [comment]
The current Git version has an incorrectly-sized icon for me under both Mac OS X and Linux on my Retina MacBook.

Changing `width: 32px;` to `width: 16px;` on line 46 of `style.css` corrects it, even when `layout.css.devPixelsPerPx` is set to other values in `about:config`.

#### Linux, width: 32px
![2015-12-22-040354_427x181_scrot](https://cloud.githubusercontent.com/assets/5890747/11947666/53c1bce6-a862-11e5-92ed-a14ec2da6a3a.png)
#### Linux, width: 16px
![2015-12-22-040319_573x171_scrot](https://cloud.githubusercontent.com/assets/5890747/11947669/5de3a2f2-a862-11e5-9329-03de1b7f06e5.png)

#### OS X, width: 32px
<img width="208" alt="screen shot 2015-12-22 at 04 06 55" src="https://cloud.githubusercontent.com/assets/5890747/11947675/6d022402-a862-11e5-9457-f4896fe6b757.png">
#### OS X, width: 16px
<img width="206" alt="screen shot 2015-12-22 at 04 07 38" src="https://cloud.githubusercontent.com/assets/5890747/11947681/7a6ea782-a862-11e5-8b62-a7ceb382e1ca.png">

**Comment by lydell** [comment]
@dwhinham Can you try if adding this at the end of style.css works?

```css
@media (min-resolution: 3dppx) {
  #VimFxButton image {
    width: 16px;
  }
}
```

**Comment by lydell** [comment]
ping @dwhinham

**Comment by dwhinham** [comment]
@lydell Apologies, Christmas and alcohol and stuff happened. :smile: 

Adding the block above does not work - I get the oversized icon again.
Cheers!

**Comment by lydell** [comment]
@dwhinham Looking at your screen shots again it does not look like you use the default Firefox theme.

- Which theme do you use?
- Does the VimFx icon get the correct size if you disable your theme (or try in a clean profile)?

**Comment by dwhinham** [comment]
Under Linux I use the GNOME Adwaita theme: https://addons.mozilla.org/en-GB/firefox/addon/adwaita/
Under OS X I use the default theme.

Creating a brand new profile with no changes/themes/extensions and adding VimFX results in exactly the same results as shown in my screenshots on both operating systems.

Again, `width: 16px;` should be the proper entry in styles.css. It seems that here, the logical size is 16 units, which given a 32-pixel icon makes sense on a double-DPI screen.

I'm not sure how the user above came to the conclusion that this was fixed earlier, but if they are playing with `layout.css.devPixelsPerPx` to make Firefox look correct, it implies that their X server's DPI settings may be incorrect. Firefox looks correct without hacks on my HiDPI Xorg setup with the proper options set in the xorg.conf.d.

**Comment by lydell** [comment]
@dwhinham I don't have a Retina screen to test on, but what you say makes a lot of sense so I've switched to 16px now.

@Hexcles When you get this update you'll need to add the following code to custom [styling](https://github.com/akhodakivskiy/VimFx/blob/master/documentation/styling.md):

```css
#VimFxButton image {
  width: 16px !important;
}
```

**Comment by dwhinham** [comment]
Many thanks :smiley: If anyone else with a Retina screen can +1 my findings it'd be appreciated.

**Comment by tll549** [comment]
Still happening on my Retina screen. (VimFx 0.10.0)

**Comment by lydell** [comment]
@tll549 The fix hasn't been released yet.

**Comment by Hexcles** [comment]
@lydell so sorry I somehow missed your comment even though you explicitly mentioned me... Should've followed this discussion. My bad.

Adding this custom style did fix all my issues: the toolbar icon shows correctly under both normal mode and hi-dpi mode.

However, after I read the [MDN link](https://developer.mozilla.org/en-US/Firefox/Australis_add-on_compat) you provided, I started to wonder why we need to specify the width of the image under 2dppx at the beginning. The MDN article basically suggests using images of different sizes (32px and 64px) under the normal mode and the double-scale mode, without explicitly specifying any rules regarding the width.

Hence, I went ahead trying the following custom style

```css
#VimFxButton image {
  width: inherit !important;
}
```

which I believe has the same effect as removing the width rule for 2dppx media and should not affect the normal 1dppx mode, since I'm essentially forcing the width to be its default value. And it turned out to work perfectly!

Did I miss anything here? Or did you have any particular reason adding that width rule under the 2dppx case? If not, probably let's try removing this rule altogether.

**Comment by Hexcles** [comment]
BTW, theoretically, at least in a normal web page instead of the browser chrome, I believe a width of 16px is what we really want, no matter what `devPixelsPerPx` we have. The whole point of the device scaling is to allow people to use "px" in HTML regardless of the device DPI, kind of making "px" in HTML a virtual unit instead of real "pixels". Therefore, we can specify sizes in "px" as if we are dealing with "normal" DPI, and the browser will automatically scale all "px" up if the device has a high DPI. However, something goes wrong in the browser chrome if we specify a width of 16px but use a 32px image, causing the icon to disappear.

Anyway, my proposal above seems to worth a try. And I found that even without a high-res monitor, one can still test the hi-dpi mode by setting `layout.css.devPixelsPerPx` to 2 (yes, you'll get a large and blurry interface).

**Comment by Hexcles** [comment]
@dwhinham there are several "DPI" numbers in Xorg. GNOME>=3.12 can automatically enable hi-dpi mode and set `Xft.dpi` to 192, but leaves another DPI setting in Xorg server intact. In such case, it is necessary to play with `devPixelsPerPx` in Firefox. Users used to do this themselves, but Firefox starts to set this value according to GNOME settings after it migrates to GTK+3 ([ArchLinux Wiki](https://wiki.archlinux.org/index.php/HiDPI#Firefox)). Anyway, it's off-topic.

**Comment by lydell** [comment]
@dwhinham Could you try out the following code as @Hexcles suggested? That’d be awesome.

```css
#VimFxButton image {
  width: inherit !important;
}
```

**Comment by Hexcles** [comment]
Hmm, so the width property under 2dppx was initially introduced to solve this exact issue. Setting the width to inherit should effectively revert everything back to the beginning (i.e. without the width property), but this time I don't see the double-sized icon. Weird. I'll find a time to hack the source directly.

**Comment by dwhinham** [comment]
@Hexcles 
I have an entry in `/etc/X11/xorg.conf.d/50-color-lcd.conf`, [as per the Arch Wiki](https://wiki.archlinux.org/index.php/xorg#Display_size_and_DPI), that does this:
```
Section "Monitor"
    Identifier             "<default monitor>"
    DisplaySize            381 238
EndSection
```
This takes care of the 'other Xorg DPI setting' you mention. I don't use GNOME, but I do set my Xft DPI manually. After these two changes to my system, Firefox (and most other apps) looks fine out of the box.

**Comment by dwhinham** [comment]
@lydell @Hexcles 

Hexcles' code works great for me for all DPIs. Thanks @Hexcles :smile: :+1:
That code also takes care of #708 for me.

**Comment by lydell** [comment]
@dwhinham Thank you so much for your help!

@Hexcles Thanks for finding the solution!

**Comment by Hexcles** [comment]
Coming back to report the result of the experiment (removing the special width rule for high DPI completely) I mentioned earlier.

I removed the whole rule for `#VimFxButton image` under the hi-dpi media query (basically _reverting_ 78c552a6d3c959a0d7f1fa9b70687b07dfafe2ac), without enforcing `inherit` either. Now in Firefox 46, everything works fine under any DPI.

Notice that my modified code is roughly the same as when I created this issue. Yet back then (I think it was Firefox 42), it would cause the toolbar icon disappear under high DPI. I suspect there was some bug in Firefox and now it's fixed (I failed to find related bug reports, however).

To sum up, the code snippet in MDN is correct and works flawlessly now. Also @lydell , feel free to remove https://github.com/akhodakivskiy/VimFx/blob/develop/extension/skin/style.css#L30-L32 to keep the code cleaner, as I've tested that this rule basically does nothing.

Really glad to see this annoying (mysterious?) issue come to an end!

**Comment by lydell** [comment]
@Hexcles are you sure it will work for @dwhinham as well?

**Comment by Hexcles** [comment]
@lydell Most likely -- I tested on my normal and hi-dpi monitors. Both were fine. Actually, I came up with the custom style to achieve the effect of removing the rule. However, I don't know if there's any easy way to test it without modifying the source and rebuilding the xpi.

If we want to be extra cautious, and if @dwhinham is so kind to help, we can remove the rule in `develop` branch and test the dev build. You may provide the (signed) xpi directly to make the testing even easier!

**Comment by lydell** [comment]
@Hexcles you’re right, there’s no need to set the `width` of the button image anymore. This was indeed a [bug in Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1208715), which was fixed in Firefox (mentioned in the [Add-on Compatibility for Firefox 44](https://blog.mozilla.org/addons/2015/12/29/compatibility-for-firefox-44/) blog post).
